### PR TITLE
Remove Kokkos Kernels Dependency

### PR DESCRIPTION
**Pull request by sslattery**
Removes the KokkosKernels dependency for now to clean up our current build issues. Recreates the needed LU and solve functionality. Will also add something later for the general eigendecomposition. Also adds a deep copy operation for batched linear algebra objects for those cases where you want to ensure the copy happens. This can be the case when using views.

The intention is to eventually go back to KokkosKernels when some issues are resolved.

### CODE REVIEW

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 1]
```diff
@@ -1696,44 +1798,104 @@ operator^( const ExpressionA& a, const ExpressionB& b )
...
+    {
+        const int iend = m - p - 1;
+        const int jend = n;
+
+        for ( int i = 0; i < iend; ++i )
+#if defined( KOKKOS_ENABLE_PRAGMA_UNROLL )
+#pragma unroll
+#endif
+            for ( int j = 0; j < jend; ++j )
+                x( i + p + 1, j ) -= a_lu( i + p + 1, p ) * x( p, j );
```
Comment (terwin) [inline src/Picasso_BatchedLinearAlgebra.hpp:1825]: No need for the +/- 1 on the `i` index:
```suggestion
        const int iend = m - p;
        const int jend = n;

        for ( int i = 1; i < iend; ++i )
#if defined( KOKKOS_ENABLE_PRAGMA_UNROLL )
#pragma unroll
#endif
            for ( int j = 0; j < jend; ++j )
                x( i + p, j ) -= a_lu( i + p, p ) * x( p, j );
```

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 2]
```diff
@@ -1092,11 +1114,66 @@ struct VectorView
...
+//---------------------------------------------------------------------------//
+template <class ExpressionA, class ExpressionB,
+          typename std::enable_if_t<is_matrix<ExpressionA>::value &&
+                                        is_matrix<ExpressionB>::value,
+                                    int> = 0>
+KOKKOS_INLINE_FUNCTION void deepCopy( ExpressionA& a, const ExpressionB& b )
+{
+    static_assert( std::is_same<typename ExpressionA::value_type,
+                                typename ExpressionB::value_type>::value,
+                   "value_type must match" );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1143]: Possibly but logic is used throughout. Let's explore in another PR

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 3]
```diff
@@ -524,8 +515,8 @@ struct Matrix<T, 1, 1>
     using eval_type = MatrixView<T, 1, 1>;
     using copy_type = Matrix<T, 1, 1>;
 
-    // Default constructor.
     KOKKOS_DEFAULTED_FUNCTION
+    // Default constructor.
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:519]: Must have been a typo - fixed

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 4]
```diff
@@ -1092,11 +1114,66 @@ struct VectorView
...
+//---------------------------------------------------------------------------//
+template <class ExpressionX, class ExpressionY,
+          typename std::enable_if_t<is_vector<ExpressionX>::value &&
+                                        is_vector<ExpressionY>::value,
+                                    int> = 0>
+KOKKOS_INLINE_FUNCTION void deepCopy( ExpressionX& x, const ExpressionY& y )
+{
+    static_assert( std::is_same<typename ExpressionX::value_type,
+                                typename ExpressionY::value_type>::value,
+                   "value_type must match" );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1167]: Possibly but logic is used throughout. Let's explore in another PR

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 5]
```diff
@@ -1237,21 +1318,21 @@ operator*( const ExpressionX& x, const ExpressionA& a )
...
+
+    for ( int i = 0; i < ExpressionX::extent_0; ++i )
+#if defined( KOKKOS_ENABLE_PRAGMA_UNROLL )
+#pragma unroll
+#endif
+        for ( int j = 0; j < ExpressionA::extent_1; ++j )
+            y( i, j ) += x_eval( i ) * a_eval( 0, j );
+
     return y;
 }
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1337]: It wasn't general before - it was explicitly for this purpose as some expressions created `Nx1` matrices. For a given vector of size `N` this production is only possible for an `Nx1` matrix.

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 6]
```diff
@@ -1469,8 +1548,32 @@ KOKKOS_INLINE_FUNCTION
                               typename ExpressionA::copy_type>
     LU( const ExpressionA& a )
 {
+    using value_type = typename ExpressionA::value_type;
+    const int m = ExpressionA::extent_0;
+    const int n = ExpressionA::extent_1;
+    const int k = ( m < n ? m : n );
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1554]: yes - fixed

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 7]
```diff
@@ -1469,8 +1548,32 @@ KOKKOS_INLINE_FUNCTION
...
+        for ( int i = 0; i < iend; ++i )
+        {
+            lu( i + p + 1, p ) /= diag;
+#if defined( KOKKOS_ENABLE_PRAGMA_UNROLL )
+#pragma unroll
+#endif
+            for ( int j = 0; j < jend; ++j )
+                lu( i + p + 1, j + p + 1 ) -=
+                    lu( i + p + 1, p ) * lu( p, j + p + 1 );
+        }
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1574]: fixed

File: src/Picasso_BatchedLinearAlgebra.hpp (hunk 1 above)
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1825]: fixed

File: src/Picasso_BatchedLinearAlgebra.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (terwin) [reply src/Picasso_BatchedLinearAlgebra.hpp:520]: I was thinking the comment belongs where it was originally, above the Kokkos macro.

File: src/Picasso_BatchedLinearAlgebra.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (terwin) [reply src/Picasso_BatchedLinearAlgebra.hpp:1338]: In that case, I think the comment and `static_assert` text could be more clear. This is not really vector-matrix multiplication, but a dot-product where the RHS vector is a `1xN` matrix. It also needs a `static_assert` on the vector length: `ExpressionX::extent_0 == ExpressionA::extent_1`.

[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp:1338]: The `+=` actually isn't necessary - I updated code and comments to reflect that fact that this is really an overload for the `x * ~y` operation (vector times vector-transpose producing a matrix).

File: src/Picasso_BatchedLinearAlgebra.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp]: whoops - fixed again

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 8]
```diff
@@ -319,7 +292,12 @@ struct VectorExpression
 
     // Evaluate the expression at an index.
     KOKKOS_INLINE_FUNCTION
-    value_type operator()( const int i ) const { return _f( i ); }
+    auto operator()( const int i ) const { return _f( i ); }
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp]: I eliminated this in favor of a larger AD exploration in the future with these expressions.

File: src/Picasso_BatchedLinearAlgebra.hpp [hunk 9]
```diff
@@ -320,12 +293,17 @@ struct VectorExpression
     // Evaluate the expression at an index.
     KOKKOS_INLINE_FUNCTION
     value_type operator()( const int i ) const { return _f( i ); }
+
+    // Evaluate the expression at an index. 2D version for vectors treated as
+    // matrices.
+    KOKKOS_INLINE_FUNCTION
+    auto operator()( const int i, int ) const { return _f( i ); }
```
Comment (terwin) [inline src/Picasso_BatchedLinearAlgebra.hpp:300]: You missed this one.

Comment (sslattery) [reply src/Picasso_BatchedLinearAlgebra.hpp]: awww man - fixed

**Review by terwin — APPROVED** [review]

**Pull request merged**